libmariadb/mariadb_lib.c

```diff
@@ -1080,6 +1161,73 @@ int mthd_my_read_one_row(MYSQL *mysql,uint fields,MYSQL_ROW row, ulong *lengths)
   return 0;
 }
 
+/**
+ * If CLIENT_DEPRECATE_EOF isn't supported by the server, the response looked like
+ * <FieldCount><Metadata><EOF><ResultSet * n><EOF>.
+ *
+ * Otherwise, the response looks like <FieldCount><Metadata><ResultSet * n><EOF>.
+ * The number of metadata to look for is already available in FieldCount
+ * therefore first EOF is not necessary.
+ *
+ * Therefore this method is introduced to just read packet metadata 
+ * and is called by both mthd_my_read_query_result and mthd_stmt_read_prepare_response
+ */
+MYSQL_FIELD *mthd_my_read_metadata_ex(MYSQL *mysql, MA_MEM_ROOT *alloc,
+                                     ulong field_count, uint field)
+{
+  ulong *len;
+  uint  f;
+  uchar *pos;
+  MYSQL_FIELD *fields, *result;
+  MYSQL_ROWS data;
+  NET *net = &mysql->net;
+
+  len= (ulong*) ma_alloc_root(alloc, sizeof(ulong)*field);
+  fields= result= (MYSQL_FIELD *) ma_alloc_root(alloc,
+      sizeof(MYSQL_FIELD)*field_count);
+
+  if (!result)
+  {
+    SET_CLIENT_ERROR(mysql, CR_OUT_OF_MEMORY, SQLSTATE_UNKNOWN, 0);
+    return(0);
+  }
+
+  data.data= (MYSQL_ROW) ma_alloc_root(alloc, sizeof(char *)*(field+1));
+  memset(data.data, 0, sizeof(char *)*(field+1));
+
+  /*
+     In this below loop we read each column info as 1 single row
+     and save it in mysql->fields array
+     */
+  for (f=0 ; f < field_count ; ++f)
+  {
+    if(mthd_my_read_one_row(mysql, field, data.data, len) == -1)
+      return NULL;
+    if(unpack_field(mysql, alloc, 0, &data, fields++))
+      return NULL;
+  }
+  if (!(mysql->server_capabilities & CLIENT_DEPRECATE_EOF))
+  {
+    if (packet_error == ma_net_safe_read(mysql, NULL))
+      return 0;
+
+    pos= net->read_pos;
+    if (*pos == 254)
+    {
+      mysql->warning_count= uint2korr(pos + 1);
+      mysql->server_status= uint2korr(pos + 3);
+    }
+  }
+
+  return result;
+}
+
+MYSQL_FIELD *mthd_my_read_metadata(MYSQL *mysql, ulong field_count, uint field)
```

Comment:
The reason mthd_my_read_metadata_ex and mthd_my_read_metadata are two methods: similar to how it is implemented in mysql client. mthd_my_read_metadata_ex takes MA_MEM_ROOT so it can be called from mthd_stmt_read_prepare_response